File: heap/heap.ml
type 'a leftist =
    | Leaf
    | Node of 'a leftist * 'a * 'a leftist * int
;;

let singleton x = Node (Leaf, x, Leaf, 1);;

let rank = function
    | Leaf -> 0
    | Node (_,_,_,r) -> r
;;

let rec merge t1 t2 =
    match t1, t2 with
    | Leaf, t | t, Leaf -> t
    | Node (l, k1, r,_), Node (_, k2, _,_) ->
        if k1>k2 then merge t2 t1
        else
            let merged = merge r t2 in
            let rank_right = rank merged and rank_left = rank l in
            if rank_right > rank_left then
                Node (merged, k1, l, rank_left+1)
            else
                Node (l, k1, merged, rank_right+1)
;;

let insert x t = merge (singleton x) t;;

let get_min = function
    | Leaf -> failwith "empty"
    | Node (_,k,_,_) -> k
;;

let delete_min = function
    | Leaf -> failwith "empty"
    | Node (l,k,r,_) -> merge l r
;;

File: heap/heap.c
#include <stdlib.h>
#include <stdio.h>
#include "heap.h"

typedef HeapNode Node;

Node* singleton(int x) {
    Node *new = malloc(sizeof *new);
    new->key = x;
    new->rank = 1;
    new->left = NULL;
    new->right = NULL;
    return new;
}

int rank(Node *head) {
    if (head == NULL)
        return 0;
    return head->rank;
}

Node* merge(Node *one, Node *two) {
    if (one == NULL)
        return two;
    if (two == NULL)
        return one;
    if (one->key>two->key)
        return merge(two, one);
    Node *merged = merge(one->right, two);
    int rank_right = rank(merged), rank_left = rank(one->left);
    if (rank_right > rank_left) {
        one->right = one->left;
        one->left = merged;
        one->rank = rank_left+1;
    }
    else {
        one->right = merged;
        one->rank = rank_right+1;
    }
    return one;
}

void free_nodes(Node **head) {
    if (head == NULL || *head == NULL)
        return;
    free_nodes(&((*head)->left));
    free_nodes(&((*head)->right));
    free(*head);
    *head=NULL;
}

void insert(MinHeap *heap, int x) {
    if (heap->head != NULL) {
        Node* new_head = merge(singleton(x), heap->head);
        heap->head = new_head;
    }
    else {
        heap->head = singleton(x);
    }
}

int get_min(MinHeap heap) {
    if (heap.head == NULL) {
        printf("Error, empty heap!\n");
        exit(EXIT_FAILURE);
    }
    return heap.head->key;
}

void delete_min(MinHeap *heap) {
    if (heap->head == NULL) { 
        printf("Error, empty heap!\n");
        exit(EXIT_FAILURE);
    }
    Node *new_head = merge(heap->head->left, heap->head->right);
    free(heap->head);
    heap->head = new_head;
}

int isempty(MinHeap heap) {
    return heap.head == NULL;
}

void clear(MinHeap *heap) {
    free_nodes(&(heap->head));
}

